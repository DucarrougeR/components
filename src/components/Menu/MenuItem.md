`MenuItem` is an `HTMLAnchorElement` that supports all of the `BoxProps` properties.

An icon can be optionally be assigned to each item via the `icon` property.

```js
<MenuItem>Home</MenuItem>
<MenuItem icon="Home">Home</MenuItem>
<MenuItem href="https://google.com" target="_blank" icon="Public">Away</MenuItem>
```

### Detail Text

Use the `detail` prop to provide explanatory detail to a given item.

```js
<MenuItem detail="Get regular updates">Schedule</MenuItem>
<MenuItem icon="Beaker" detail="Try the newest toys!">
  Labs
</MenuItem>
```

### Active & MenuGroup canActivate

Use the `active` prop to indicate the selected item(s).

Note: This property behaves slightly diffently depending on the context of the `MenuGroup` it is contained within. If the `MenuGroup` has the `canActivate` prop it will reserve space for the checkmark next to `active` items and any `icon` assignments on the items will be ignored.

```js
<Menu>
  <MenuGroup
    label="canActivate=false, no icons"
    labelProps={{ transform: 'none', bg: 'transparent' }}
  >
    <MenuItem active>Gouda</MenuItem>
    <MenuItem detail="Is often orange">Cheddar</MenuItem>
    <MenuItem>Swiss</MenuItem>
  </MenuGroup>
  <MenuGroup
    label="canActivate=false, icons"
    labelProps={{ transform: 'none', bg: 'transparent' }}
  >
    <MenuItem icon="Favorite" active>
      Gouda
    </MenuItem>
    <MenuItem icon="Favorite" detail="Is often orange">
      Cheddar
    </MenuItem>
    <MenuItem icon="Favorite">Swiss</MenuItem>
  </MenuGroup>
  <MenuGroup
    canActivate
    label="canActivate=true, icon ignored"
    labelProps={{ transform: 'none', bg: 'transparent' }}
  >
    <MenuItem active>Gouda</MenuItem>
    <MenuItem icon="Favorite" detail="Is often orange">
      Cheddar
    </MenuItem>
    <MenuItem>Swiss</MenuItem>
  </MenuGroup>
</Menu>
```


### Digital Marketing Menu
```js
const palette = require('../../style').palette

const menuCustomizations =
  {
    backgroundColor: palette.charcoal100,
    backgroundColorHover: palette.charcoal100,
    backgroundColorCurrent: palette.charcoal200,
    textColor: palette.charcoal800,
    textColor: palette.charcoal800,
    textColorCurrent: palette.charcoal900,
    iconColor: palette.charcoal800,
    iconColorHover: palette.purple500,
    iconColorCurrent: palette.purple500,
    iconSize: 20,
    currentMarkerColor: palette.purple300 }
  ;

  <MenuGroup label="GOOGLE ADS">
    <MenuItem icon="Home"  customizableProps={menuCustomizations}>
        Gouda
      </MenuItem>
      <MenuItem icon="FavoriteOutline" current detail="Is often orange" customizableProps={menuCustomizations}>
        Cheddar
      </MenuItem>
    <MenuItem icon="Dashboard" customizableProps={menuCustomizations}>Swiss</MenuItem>
  </MenuGroup>

```

### Customized menu

```js
const palette = require('../../style').palette

const menuCustomizations =
  {
    backgroundColor: palette.purple500,
    backgroundColorHover: palette.purple700,
    backgroundColorCurrent: palette.purple200,
    textColor: palette.purple200,
    textColorHover: 'white',
    textColorCurrent: palette.purple900,
    iconColor: palette.purple300,
    iconColorHover: palette.purple100,
    iconColorCurrent: palette.purple500,
    iconSize: 20,
    currentMarkerColor: palette.purple300 }
  ;

  <MenuGroup>
    <MenuItem icon="Home"  customizableProps={menuCustomizations}>
        Gouda
      </MenuItem>
      <MenuItem icon="FavoriteOutline" current  currentMarker  detail="Is often orange" customizableProps={menuCustomizations}>
        Cheddar
      </MenuItem>
    <MenuItem icon="Dashboard" customizableProps={menuCustomizations}>Swiss</MenuItem>
  </MenuGroup>

```
