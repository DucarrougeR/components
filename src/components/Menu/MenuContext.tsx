import * as React from 'react'
import { MenuItemCustomization } from './MenuItem/menuItemCustomization'

export interface MenuContextProps {
  customizationProps?: MenuItemCustomization
}

export const MenuContext = React.createContext<MenuContextProps>({})

export function withMenu<T extends MenuContextProps>(
  Component: React.ComponentType<T>
) {
  return function wrapContext(props: T) {
    return (
      <MenuContext.Consumer>
        {contextProps => <Component {...props} {...contextProps} />}
      </MenuContext.Consumer>
    )
  }
}
