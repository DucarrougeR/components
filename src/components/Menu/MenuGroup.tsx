import * as React from 'react'
import { palette, styled } from '../../style'
import { Box, BoxPropsWithout } from '../Box'
import { Heading, HeadingProps } from '../Heading'
import { List } from '../List'
import { MenuContext } from './MenuContext'
import { MenuItemCustomization } from './MenuItem'

export interface MenuGroupProps
  extends BoxPropsWithout<HTMLDivElement, 'label'> {
  label?: React.ReactNode
  labelProps?: HeadingProps
  labelStyles?: React.CSSProperties
  customizationProps?: MenuItemCustomization
  compact?: boolean
  groupLabelShadow?: boolean
}

const InternalMenuGroup: React.FC<MenuGroupProps> = ({
  groupLabelShadow,
  children,
  label,
  labelProps,
  labelStyles,
  ...props
}) => {
  const { customizationProps, compact, ...boxProps } = props
  const menu = React.useContext(MenuContext)

  const customizations = customizationProps
    ? customizationProps
    : menu.customizationProps

  const shadow =
    groupLabelShadow !== undefined ? groupLabelShadow : menu.groupLabelShadow

  const labelComponent = label && (
    <Heading
      bg={customizations ? customizations.bg : palette.white}
      fontSize="xsmall"
      is="h2"
      px="medium"
      py={compact || menu.compact ? 'small' : 'xxsmall'}
      transform="upper"
      position={shadow ? 'sticky' : 'initial'}
      top="0"
      fontWeight="semiBold"
      boxShadow={shadow ? `0 4px 8px -2px ${palette.charcoal200}` : 'none'}
      {...labelProps}
      style={labelStyles}
      zIndex={2}
    >
      {label}
    </Heading>
  )

  return (
    <MenuContext.Provider
      value={{
        compact: compact || menu.compact,
        customizationProps: customizationProps || menu.customizationProps,
      }}
    >
      <Box
        is="li"
        {...boxProps}
        bg={customizations ? customizations.bg : palette.white}
        py="xxsmall"
      >
        {labelComponent}
        <List nomarker>{children}</List>
      </Box>
    </MenuContext.Provider>
  )
}

const MenuGroupFactory = React.forwardRef((props: MenuGroupProps, ref) => (
  <InternalMenuGroup innerRef={ref} {...props} />
))

export const MenuGroup = styled(MenuGroupFactory)``
