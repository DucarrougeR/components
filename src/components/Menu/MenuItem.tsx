import deepmerge from 'deepmerge'
import * as React from 'react'
import { IconNames } from '../../icons/build/IconNames'
import {
  css,
  easings,
  FontWeights,
  palette,
  styled,
  transitions,
} from '../../style'
import { Box, BoxProps, BoxPropsWithout } from '../Box'
import { Icon } from '../Icon'
import { MenuContextProps, withMenu } from './MenuContext'

export interface MenuMarkerCustomizations {
  color?: string
  size?: number
}

export interface MenuIconCustomizations {
  color?: string
  size?: number
}

export interface MenuInteractiveCustomizations {
  bg?: string
  color?: string
  text?: string
  icon?: MenuIconCustomizations
  fontWeight?: FontWeights
}

export interface MenuItemCustomizationProps extends BoxProps<HTMLDivElement> {
  bg: string
  color: string
  marker: MenuMarkerCustomizations
  icon: MenuIconCustomizations
  hover: MenuInteractiveCustomizations
  current: MenuInteractiveCustomizations
  activated: MenuInteractiveCustomizations
}

export interface MenuIconProps
  extends BoxPropsWithout<HTMLDivElement, 'name' | 'color' | 'size'> {
  color: string
  size: number
  hoverColor: string
  currentColor: string
  activeColor: string
}

export interface MenuItemProps
  extends BoxProps<HTMLAnchorElement>,
    MenuContextProps {
  detail?: React.ReactNode
  icon?: IconNames
  /**
   * Indicates the MenuItem is checked
   */
  active?: boolean
  /**
   * Display the MenuItem with a checklist style
   * - Use with `itemRole='button'`
   */
  canActivate?: boolean
  /**
   * Indicates the MenuItem represents the currently displayed page
   * - Use with `itemRole='link'`
   */
  current?: boolean
  /**
   * Display a marker next to the MenuItem if it is current
   */
  currentMarker?: boolean
  /**
   * Custiomize the appearance of the MenuItem
   */
  customizationProps?: MenuItemCustomizationProps
  /**
   * Sets the correct accessible role for the MenuItem:
   * - Use **'link'** for items that navigation to another page
   * - Use **'button'** for items that trigger in page interactions, like displaying a modal
   * @default 'button'
   *
   */
  itemRole?: 'link' | 'button'
  onClick?: () => void
}

const MenuItemInteral: React.FC<MenuItemProps> = ({
  active,
  current,
  currentMarker,
  canActivate,
  children,
  detail,
  icon,
  customizationProps,
  onClick,
  itemRole,
  ...props
}) => {
  // tslint:disable:object-literal-sort-keys
  const defaultcustomizationProps: MenuItemCustomizationProps = {
    bg: palette.white,
    color: palette.charcoal600,
    icon: {
      color: palette.charcoal300,
      size: 20,
    },
    marker: {
      size: 4,
      color: palette.charcoal900,
    },
    hover: {
      bg: palette.charcoal100,
      color: palette.charcoal900,
      icon: {
        color: palette.charcoal900,
      },
    },
    current: {
      bg: palette.charcoal100,
      color: palette.charcoal900,
      fontWeight: 'bold',
      icon: {
        color: palette.charcoal900,
      },
    },
    activated: {
      color: palette.blue500,
      icon: {
        color: palette.blue500,
      },
    },
  }
  // tslint:enable:object-literal-sort-keys

  const customProps: MenuItemCustomizationProps = customizationProps
    ? deepmerge(defaultcustomizationProps, customizationProps)
    : defaultcustomizationProps

  const formatDetail = (content?: React.ReactNode) =>
    content ? (
      <Box
        pl="large"
        ml="auto"
        fontSize="xsmall"
        color={palette.charcoal300}
        zIndex={1}
      >
        {content}
      </Box>
    ) : null

  const itemIcon = () => {
    const placeholder = <Box width="1.5rem" />
    const iconComponent = (name: IconNames) => (
      <MenuIcon name={name} mr="xsmall" size={customProps.icon.size} />
    )

    if (canActivate) {
      return active ? iconComponent('Check') : placeholder
    } else if (icon) {
      return iconComponent(icon)
    } else {
      return
    }
  }
  const childrenWrapper = () => {
    return (
      <MenuItemChildStyle
        is="button"
        role={canActivate ? 'menuitemradio' : 'menuitem'}
        tabIndex={itemRole === 'link' ? -1 : undefined}
        aria-checked={active}
      >
        {itemIcon()}
        {children}
      </MenuItemChildStyle>
    )
  }

  return (
    <MenuItemStyle
      is="li"
      alignItems="center"
      color={active ? customProps.activated.color : customProps.color}
      display="flex"
      flexWrap="wrap"
      fontSize="small"
      py="small"
      px="medium"
      onClick={onClick}
      bg={customProps.bg}
      active={active}
      activeStyle={{ color: customProps.activated.color }}
      current={current}
      aria-current={current && 'page'}
      currentMarker={currentMarker}
      customizationProps={customProps}
      {...props}
    >
      {childrenWrapper()}
      {formatDetail(detail)}
    </MenuItemStyle>
  )
}

function currentStyles(props: StyleProps) {
  if (props.current) {
    return `
      background: ${props.customizationProps.current.bg};
      color: ${props.customizationProps.current.color};
      font-weight: ${props.customizationProps.current.fontWeight}
    `
  }
  return false
}

function currentBorder(props: StyleProps) {
  if (props.current && props.currentMarker) {
    return css`
      ::before {
        content: '';
        display: block;
        height: 100%;
        position: absolute;
        left: 0;
        top: 0;
        background: ${props.customizationProps.marker.color};
        width: ${props.customizationProps.marker.size}px;
      }
    `
  }

  return false
}

function iconColor(props: StyleProps) {
  if (props.active) {
    return css`
      ${MenuIcon} {
        color: ${props.customizationProps.activated.icon!.color};
      }
    `
  } else if (props.current) {
    return css`
      ${MenuIcon} {
        color: ${props.customizationProps.current.icon!.color};
      }
    `
  } else {
    return css`
      ${MenuIcon} {
        color: ${props.customizationProps.icon.color};
      }
    `
  }
}

function hoverStyles(props: StyleProps) {
  if (props.current) {
    return false
  } else {
    return css`
      :hover {
        background: ${props.customizationProps.hover.bg};
        color: ${props.customizationProps.hover.color};

        ${MenuIcon} {
          color: ${props.customizationProps.hover.icon &&
            props.customizationProps.hover.icon.color};
        }
      }
    `
  }
}

interface StyleProps extends MenuItemProps {
  customizationProps: MenuItemCustomizationProps
}

//
// All of this  drama is to not auto-spread bad props onto Box and cause React run-time warnings
//
const MenuItemStyleFactory = (props: StyleProps) => {
  const {
    active,
    current,
    currentMarker,
    customizationProps,
    ...boxProps
  } = props
  return <Box {...boxProps} />
}

const MenuIcon = styled(Icon)``

const MenuItemStyle = styled(MenuItemStyleFactory)`
  position: relative;
  text-decoration: none;
  transition: background ${transitions.durationQuick} ${easings.ease},
    color ${transitions.durationQuick} ${easings.ease};
  ${hoverStyles};
  ${MenuIcon} {
    transition: color ${transitions.durationQuick} ${easings.ease};
  }
  ${iconColor};
  ${currentStyles};
  ${currentBorder};

  :focus-within {
    box-shadow: 0 0 0.25rem 0.125rem ${palette.blue400};
    background: ${props => !props.current && props.customizationProps.hover.bg};
    outline: 'none';
    z-index: 1;
  }
`

const MenuItemChildStyleFactory = (props: MenuItemProps) => {
  return <Box border="none" {...props} />
}

const MenuItemChildStyle = styled(MenuItemChildStyleFactory)`
  outline: none;
  background: transparent;
  color: inherit;
  text-align: left;
  flex: 1;
  font-size: inherit;
  display: flex;
  align-items: center;
  cursor: pointer;
`

export const MenuItem = withMenu(MenuItemInteral)
