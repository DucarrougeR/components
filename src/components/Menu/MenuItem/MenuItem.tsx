import deepmerge from 'deepmerge'
import * as React from 'react'
import { IconNames } from '../../../icons/build/IconNames'
import { RampSizes, styled } from '../../../style'
import { BoxProps } from '../../Box'
import { Icon } from '../../Icon'
import { MenuContext, MenuContextProps } from '../MenuContext'
import { MenuItemButton } from './MenuItemButton'
import {
  MenuItemCustomization,
  MenuItemStateCustomizations,
} from './menuItemCustomization'
import { MenuItemDetail } from './MenuItemDetail'
import { MenuItemListItem } from './MenuItemListItem'
import {
  defaultMenuItemStyle,
  MenuItemStateStyle,
  MenuItemStyle,
} from './menuItemStyle'

export interface MenuItemProps
  extends BoxProps<HTMLAnchorElement>,
    MenuContextProps {
  detail?: React.ReactNode
  icon?: IconNames
  /**
   * Indicates the MenuItem is checked
   */
  current?: boolean
  /**
   * Display a marker next to the MenuItem if it is current
   */
  currentMarker?: boolean
  /**
   * Sets the correct accessible role for the MenuItem:
   * - Use **'link'** for items that navigation to another page
   * - Use **'button'** for items that trigger in page interactions, like displaying a modal
   * @default 'button'
   *
   */
  itemRole?: 'link' | 'button'
  onClick?: () => void
  href?: string
  target?: string
}

const InternalMenuItem: React.FC<MenuItemProps> = props => {
  const {
    current,
    currentMarker,
    children,
    detail,
    icon,
    customizationProps,
    onClick,
    itemRole,
    href,
    target,
    compact,
    ...boxProps
  } = props

  const menu = React.useContext(MenuContext)

  let customizations = customizationProps
    ? customizationProps
    : menu.customizationProps

  if (customizationProps && menu.customizationProps) {
    customizations = deepmerge(customizationProps, menu.customizationProps)
  }

  const isCompact = compact !== undefined ? compact : menu.compact
  const compactIconModifier = isCompact ? 1.25 : 1

  const style = assignCustomizations(defaultMenuItemStyle, customizations)
  const state = current ? style.current : style.initial

  const { iconSize, iconColor, ...listItemProps } = state

  return (
    <MenuItemListItem
      current={current}
      currentMarker={currentMarker}
      itemStyle={style}
      alignItems="center"
      aria-current={current && 'page'}
      display="flex"
      flexWrap="wrap"
      is="li"
      onClick={onClick}
      {...listItemProps}
      {...boxProps}
    >
      <MenuItemButton
        is={itemRole === 'link' ? 'a' : 'button'}
        role="menuitem"
        href={href}
        target={target}
        px="medium"
        py={isCompact ? 'xxsmall' : 'small'}
      >
        {icon && (
          <Icon
            name={icon}
            mr="xsmall"
            size={iconSize / compactIconModifier}
            color={iconColor}
          />
        )}
        {children}
      </MenuItemButton>
      {detail && <MenuItemDetail>{detail}</MenuItemDetail>}
    </MenuItemListItem>
  )
}

const MenuItemFactory = React.forwardRef((props: MenuItemProps, ref) => (
  <InternalMenuItem innerRef={ref as React.RefObject<HTMLElement>} {...props} />
))

export const MenuItem = styled<MenuItemProps>(MenuItemFactory)``

const assignCustomizations = (
  defaultStyle: MenuItemStyle,
  changes?: MenuItemCustomization
): MenuItemStyle => {
  const { bg, color, iconColor, fontWeight, fontSize, iconSize } =
    changes || ({} as MenuItemCustomization)

  const customMarker = changes ? changes.marker : {}

  // Need to spread fontSize & iconSize across all states
  const defaults = {
    fontSize: fontSize || ('small' as RampSizes),
    iconSize: iconSize || 20,
  }

  const base: MenuItemStateCustomizations = {}

  bg && (base.bg = bg)
  color && (base.color = color)
  fontWeight && (base.fontWeight = fontWeight)
  iconColor && (base.iconColor = iconColor)

  const initial: MenuItemStateStyle = {
    ...defaultStyle.initial,
    ...defaults,
    ...base,
  }

  const current: MenuItemStateStyle = {
    ...defaultStyle.current,
    ...defaults,
    ...base,
    ...(changes ? changes.current : {}),
  }

  const hover: MenuItemStateStyle = {
    ...defaultStyle.hover,
    ...defaults,
    ...base,
    ...(changes ? changes.hover : {}),
  }

  const marker = { ...defaultStyle.marker, ...customMarker }

  return {
    current,
    hover,
    initial,
    marker,
  }
}
