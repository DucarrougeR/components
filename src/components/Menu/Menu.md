```js noeditor
const StatusAndResources = require('../../../styleguide_components/StatusAndResources')
  .StatusAndResources
;<StatusAndResources
  status="experimental"
  figmaURL="https://www.figma.com/file/VyHO1Hv1XaW1v3lE9I4PVT/Menu"
  githubURL="https://github.com/looker/lens/blob/master/src/components/Menu/Menu.tsx"
  feedbackTitle="Menu Component Feedback"
/>
```

`Menu` displays a list of choices.

#### Focus & Keyboard Controls

When a menu is focused the `up` and `down` arrow keys will move focus through the menu items within the menu.

```js
import { Menu } from './Menu'
import { MenuItem } from './MenuItem'
;<Menu>
  <MenuItem onClick={() => alert('Hello world!')}>Gouda</MenuItem>
  <MenuItem>Cheedar</MenuItem>
  <MenuItem>Swiss</MenuItem>
</Menu>
```

### Icon Support

```js
import { Menu } from './Menu'
import { MenuItem } from './MenuItem'
;<Menu>
  <MenuItem icon="LogoRings">Looker</MenuItem>
  <MenuItem icon="Validate">Validate</MenuItem>
  <MenuItem icon="ChartPie">Pizza</MenuItem>
</Menu>
```

### Menu Item Spacing

Menus accept a `compact` prop that will make the spacing between the `MenuItems` smaller

```js
import { Menu } from './Menu'
import { MenuItem } from './MenuItem'
;<Menu compact>
  <MenuItem icon="LogoRings">Looker</MenuItem>
  <MenuItem icon="Validate">Validate</MenuItem>
  <MenuItem icon="ChartPie">Pizza</MenuItem>
</Menu>
```


### Compose Menus

Menu can be easily composed with `Popover`, `Drawer` and `Dialog`

```js
import { Popover } from '../Overlay'
import { Button } from '../Button'
import { Menu } from './Menu'
import { MenuItem } from './MenuItem'
;<Popover
  content={
    <Menu>
      <MenuItem icon="FavoriteOutline">Gouda</MenuItem>
      <MenuItem icon="FavoriteOutline" detail="Is often orange">
        Cheddar
      </MenuItem>
      <MenuItem icon="FavoriteOutline">Swiss</MenuItem>
    </Menu>
  }
>
  {(onClick, ref) => (
    <Button
      aria-haspopup="true"
      variant="outline"
      iconAfter="ArrowDropDown"
      onClick={onClick}
      innerRef={ref}
    >
      Popover + Menu: Cheese Selector
    </Button>
  )}
</Popover>
```

```js
import { Popover } from '../Overlay'
import { Button } from '../Button'

const MenuGroupDemo = require('../../../styleguide_components/MenuDemo')
  .MenuGroupDemo

;<Popover content={<MenuGroupDemo />}>
  {(onClick, ref) => (
    <Button
      aria-haspopup="true"
      variant="outline"
      iconAfter="ArrowDropDown"
      onClick={onClick}
      innerRef={ref}
    >
      Menu Example...
    </Button>
  )}
</Popover>
```

```js
import { DrawerManager, ModalContent, ModalHeader } from '../Modal'
import { Button } from '../Button'
import { Menu } from './Menu'
import { MenuGroup } from './MenuGroup'
import { MenuItem } from './MenuItem'

const menuGroup = (
  <MenuGroup label="Spacing">
    <MenuItem>Scroll Filler A</MenuItem>
    <MenuItem>Scroll Filler B</MenuItem>
    <MenuItem>Scroll Filler C</MenuItem>
    <MenuItem>Scroll Filler D</MenuItem>
    <MenuItem>Scroll Filler E</MenuItem>
    <MenuItem>Scroll Filler F</MenuItem>
  </MenuGroup>
)
;<DrawerManager
  content={
    <>
      <ModalHeader>Menu Demo</ModalHeader>
      <ModalContent innerProps={{ p: 'none' }}>
        <Menu>
          <MenuGroup>
            <MenuItem icon="LogoRings">Looker</MenuItem>
            <MenuItem icon="Validate">Validate</MenuItem>
            <MenuItem icon="ChartPie">Pizza!</MenuItem>
          </MenuGroup>
          {menuGroup}
          {menuGroup}
          {menuGroup}
          {menuGroup}
          {menuGroup}
          <MenuGroup>
            <MenuItem icon="Beaker">Scary Stuff</MenuItem>
          </MenuGroup>
        </Menu>
      </ModalContent>
    </>
  }
>
  {onClick => (
    <Button variant="outline" onClick={onClick} iconAfter="ArrowDropDown">
      Drawer + Menu...
    </Button>
  )}
</DrawerManager>
```
