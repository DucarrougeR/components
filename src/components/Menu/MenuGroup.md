`MenuGroup` is a way to group together similar collections of `MenuItem`s.

It supports an optional `label` prop to place a heading above the `MenuItem`s. When placed within a scrolling container the label will stick to the top edge while the MenuGroup is visible. Look at the "Menu + Drawer" example above and try scrolling.

`labelProps` and `labelStyles` can be used to apply styling to the label.

```js
import { Menu } from './Menu'
import { MenuGroup } from './MenuGroup'
import { MenuItem } from './MenuItem'
;<Menu groupDividers={false}>
  <MenuGroup>
    <MenuItem icon="LogoRings">Looker</MenuItem>
    <MenuItem icon="Validate">Validate</MenuItem>
    <MenuItem icon="ChartPie">Pizza!</MenuItem>
  </MenuGroup>
  <MenuGroup label="Cheeses">
    <MenuItem icon="FavoriteOutline">Cheddar</MenuItem>
    <MenuItem icon="FavoriteOutline">Cheddar</MenuItem>
    <MenuItem icon="FavoriteOutline">Swiss</MenuItem>
  </MenuGroup>
  <MenuGroup label="Cheeses">
    <MenuItem icon="FavoriteOutline">Cheddar</MenuItem>
    <MenuItem icon="FavoriteOutline">Cheddar</MenuItem>
    <MenuItem icon="FavoriteOutline">Swiss</MenuItem>
  </MenuGroup>
</Menu>
```

You can disable the shadows for `MenuGroup` `label`'s by setting the  `groupLabelShadow` prop to false either at the `Menu` level or individually on each `MenuGroup`

```js
import { Menu } from './Menu'
import { MenuGroup } from './MenuGroup'
import { MenuItem } from './MenuItem'
;<Menu groupLabelShadow={false}>
  <MenuGroup>
    <MenuItem icon="LogoRings">Looker</MenuItem>
    <MenuItem icon="Validate">Validate</MenuItem>
    <MenuItem icon="ChartPie">Pizza!</MenuItem>
  </MenuGroup>
  <MenuGroup label="Cheeses">
    <MenuItem icon="FavoriteOutline">Cheddar</MenuItem>
    <MenuItem icon="FavoriteOutline">Cheddar</MenuItem>
    <MenuItem icon="FavoriteOutline">Swiss</MenuItem>
  </MenuGroup>
  <MenuGroup label="Cheeses">
    <MenuItem icon="FavoriteOutline">Cheddar</MenuItem>
    <MenuItem icon="FavoriteOutline">Cheddar</MenuItem>
    <MenuItem icon="FavoriteOutline">Swiss</MenuItem>
  </MenuGroup>
</Menu>
