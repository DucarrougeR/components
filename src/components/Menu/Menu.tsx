import * as React from 'react'
import { HotKeys } from 'react-hotkeys'
import { css, palette, styled } from '../../style'
import { Box, BoxProps } from '../Box'
import { MenuContext, MenuContextProps } from './MenuContext'
import { MenuGroup } from './MenuGroup'
import { moveFocus } from './moveFocus'

export interface MenuProps extends BoxProps<HTMLDivElement>, MenuContextProps {
  groupDividers?: boolean
  groupLabelShadow?: boolean
  compact?: boolean
}

const InternalMenu: React.FC<MenuProps> = ({
  customizationProps,
  groupLabelShadow = true,
  compact,
  children,
  ...props
}) => {
  const { groupDividers, ...boxProps } = props
  const ref = React.useRef<null | HTMLElement>(null)
  const keyMap = { moveDown: 'down', moveUp: 'up' }
  const keyHandlers = {
    moveDown: () => moveFocus(1, 0, ref),
    moveUp: () => moveFocus(-1, -1, ref),
  }

  return (
    <MenuContext.Provider
      value={{ customizationProps, groupLabelShadow, compact }}
    >
      <HotKeys keyMap={keyMap} handlers={keyHandlers}>
        <div ref={ref as React.RefObject<HTMLDivElement>}>
          <Box
            is="ul"
            tabIndex={-1}
            role="menu"
            userSelect="none"
            {...boxProps}
          >
            {children}
          </Box>
        </div>
      </HotKeys>
    </MenuContext.Provider>
  )
}

const MenuFactory = React.forwardRef((props: MenuProps, ref) => (
  <InternalMenu innerRef={ref} {...props} />
))

const dividersStyle = css`
  ${MenuGroup} ~ ${MenuGroup} { /* stylelint-disable-line */
    border-top: 1px solid ${palette.charcoal200};
  }
`

export const Menu = styled<MenuProps>(MenuFactory)`
  list-style: none;
  outline: none;
  ${props => props.groupDividers !== false && dividersStyle};
`
