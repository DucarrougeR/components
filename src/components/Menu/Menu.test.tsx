import { shallow } from 'enzyme'
import 'jest-styled-components'
import * as React from 'react'

import { Menu } from './Menu'
import { MenuGroup } from './MenuGroup'
import { MenuItem } from './MenuItem'

test('Menu', () => {
  const menu = shallow(
    <Menu>
      <MenuItem>boo!</MenuItem>
    </Menu>
  )

  expect(menu).toMatchSnapshot()

  expect(menu.find('div').find('a'))
})

test('Menu - focusOnMount', () => {
  const menu = shallow(
    <Menu focusOnMount>
      <MenuItem>boo!</MenuItem>
    </Menu>
  )

  expect(menu).toMatchSnapshot()

  expect(menu.find('div').find('a'))
})

test('Menu - composed', () => {
  const menu = shallow(
    <Menu>
      <MenuGroup>
        <MenuItem icon="LogoRings">Looker</MenuItem>
        <MenuItem icon="Validate">Validate</MenuItem>
        <MenuItem icon="ChartPie">Pizza!</MenuItem>
      </MenuGroup>
      <MenuGroup canActivate label="Cheeses">
        <MenuItem>Gouda</MenuItem>
        <MenuItem>Cheddar</MenuItem>
        <MenuItem>Swiss</MenuItem>
      </MenuGroup>
      <MenuGroup>
        <MenuItem icon="Beaker">Scary Stuff</MenuItem>
      </MenuGroup>
    </Menu>
  )

  expect(menu).toMatchSnapshot()
})
