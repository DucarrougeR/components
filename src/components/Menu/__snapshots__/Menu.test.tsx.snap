// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu - compact 1`] = `ReactWrapper {}`;

exports[`Menu - composed 1`] = `ShallowWrapper {}`;

exports[`Menu - menu customization 1`] = `ShallowWrapper {}`;

exports[`Menu - no label shadow 1`] = `ReactWrapper {}`;

exports[`Menu 1`] = `ShallowWrapper {}`;
